x-pack/plugins/monitoring/server/lib/alerts/fetch_cpu_usage_node_stats.ts

```diff
@@ -5,139 +5,334 @@
+    unknown,
+    typeof params
+  >;
+
+  if (!response.aggregations) {
+    throw new Error('Failed to resolve needed aggregations for CPU Usage Rule');
+  }
+
+  return response.aggregations.clusters.buckets.flatMap((cluster) => {
+    return cluster.nodes.buckets.map((node): AlertCpuUsageNodeStats => {
+      let nodeName;
+      if (node.name.buckets.length) {
+        nodeName = node.name.buckets[0].key as string;
+      }
+
+      let ccs;
+      if (node.index.buckets.length) {
+        const index = node.index.buckets[0].key as string;
+        ccs = index.includes(':') ? index.split(':')[0] : undefined;
+      }
+
+      const limitsNotSet = node.quota_micros_max.value === -1 && node.quota_micros_min.value === -1;
+      const notRunningInAContainer =
+        node.quota_micros_min.value === null && node.quota_micros_max.value === null;
+      if (limitsNotSet || notRunningInAContainer) {
+        return {
+          missingLimits: true,
+          clusterUuid: cluster.key as string,
+          nodeId: node.key as string,
+          cpuUsage: node.average_cpu_usage_percent.value ?? undefined,
+          nodeName,
+          ccs,
+        };
+      }
+
+      if (node.quota_micros_min.value !== node.quota_micros_max.value) {
+        return {
+          limitsChanged: true,
+          clusterUuid: cluster.key as string,
+          nodeId: node.key as string,
+          cpuUsage: undefined,
+          nodeName,
+          ccs,
+        };
+      }
+
+      if (
+        node.max_usage_nanos.value === null ||
+        node.min_usage_nanos.value === null ||
+        node.max_periods.value === null ||
+        node.min_periods.value === null ||
+        node.quota_micros_max.value === null
+      ) {
+        logger.warn(
+          `CPU usage rule: Some aggregated values needed for container CPU usage calculation was empty: ${findEmptyValues(
+            {
+              'max of node_stats.os.cgroup.cpu.cfs_quota_micros': node.quota_micros_max.value,
+              'max of node_stats.os.cgroup.cpuacct.usage_nanos': node.max_usage_nanos.value,
+              'min of node_stats.os.cgroup.cpuacct.usage_nanos': node.min_usage_nanos.value,
+              'max of node_stats.os.cgroup.cpu.stat.number_of_elapsed_periods':
+                node.max_periods.value,
+              'min of node_stats.os.cgroup.cpu.stat.number_of_elapsed_periods':
+                node.min_periods.value,
+            }
+          )}`
+        );
+
+        return {
+          clusterUuid: cluster.key as string,
+          nodeId: node.key as string,
+          cpuUsage: undefined,
+          nodeName,
+          ccs,
+        };
+      }
+
+      const usageNanos = node.max_usage_nanos.value - node.min_usage_nanos.value;
+      const periods = node.max_periods.value - node.min_periods.value;
+
+      const cpuUsage = (usageNanos / (node.quota_micros_max.value * periods * 1000)) * 100;
```

Comment:
**nit:** the PR description is pretty useful to understand this computation, maybe would be good to abstract it into a small function and comment it with the same description in the PR (a small portion to explain the formula) so that would be easier to understand what it does when navigating the codebase in future. Wdyt?

Reply:
I moved it to a function and added a link to the PR to get the full description.